File: 101.symmetric-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {
    return root == nil || eq(root.Left, root.Right)
}

func eq(root1, root2 *TreeNode) bool {
    if root1 == nil || root2 == nil {
        return root1 == root2
    }
    return root1.Val == root2.Val &&
        eq(root1.Left, root2.Right) &&
        eq(root1.Right, root2.Left)
}

File: 101.symmetric-tree/solution.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(var _value: Int) {
 *   var value: Int = _value
 *   var left: TreeNode = null
 *   var right: TreeNode = null
 * }
 */
object Solution {
    def isSymmetric(root: TreeNode): Boolean = {
        root == null || eq(root.left, root.right) 
    }
    
    def eq(p: TreeNode, q: TreeNode): Boolean = {
        if (p == null && q == null) return true
        if (p == null || q == null) return false
        (p.value == q.value) && eq(p.left, q.right) && eq(p.right, q.left)
    }
}